File: point_mass_vibration/mass_damper_numerical_ij.py
import numpy as np
import scipy as sp
import matplotlib.pyplot as plt

# first doing it in python and then converting to julia

# define domain and mesh
t_start = 0
t_end = 5
n_intervals = 100000
h = (t_end-t_start) / n_intervals
t_domain = np.linspace(t_start, t_end-h, n_intervals)

# define constants in si units
m = 1
c = 0.2
k = 5

# define system matrix
R = m/(h*h) + c/h + k
S = -2*m/(h*h) - c/h
T = m/(h*h)

A = sp.sparse.diags([T, S, R], [-2, -1, 0], shape=(n_intervals, n_intervals)).todense()
b = np.zeros(n_intervals)
b[0] = -1*(T+S)
b[1] = -1*T

# solve
x = np.linalg.solve(A, b)

# plot
plt.figure()
plt.plot(t_domain, x)
plt.grid(True)
plt.show()

File: point_mass_vibration/mass_damper_numerical_ij.jl
using LinearAlgebra
using SparseArrays
using Plots

# define domain and mesh
t_start = 0.0
t_end = 50.0
n_intervals = 30000
h = (t_end - t_start) / n_intervals
t_domain = range(t_start, stop=t_end - h, length=n_intervals)

# sefine constants in SI units
m = 1.0
c = 0.2
k = 5.0

# define system matrix coefficients
R = m / (h^2) + c / h + k
S = -2 * m / (h^2) - c / h
T = m / (h^2)

# create system matrix
A = spdiagm(-2 => fill(T, n_intervals-2),
            -1 => fill(S, n_intervals-1),
             0 => fill(R, n_intervals))

# define vector b
b = zeros(n_intervals)
b[1] = -1 * (T + S)
b[2] = -1 * T

# solve the system
x = A \ b

# plot the solution
plot(t_domain, x, label="Displacement", xlabel="Time (s)", ylabel="Displacement (m)", grid=true)
